astropy/nddata/mixins/ndio.py

```diff
@@ -37,3 +42,147 @@ def write(self, *args, **kwargs):
         formats.
         """
         io_registry.write(self, *args, **kwargs)
+
+
+def read_from_fits(filename):
+    # Variables (can be changed if another behaviour is desired)
+    ext_data = 0
+    ext_meta = 0
+    ext_mask = 'mask'
+    ext_uncert = 'uncert'
+    kw_unit = 'bunit'
+    kw_mask_type = 'boolmask'
+    kw_uncert_type = 'stddev'
+    # TODO: memmap had caused some problems in the tests. If memmap is used
+    # the tests checking for open-files are triggering Errors. This may be
+    # because there is a quasi-cyclic reference structure between NDData and
+    # uncertainty (if given) because the uncertainty saves a reference to
+    # the nddata in it's parent_nddata property. But that's mostly guessing at
+    # least for now memmap is False by default.
+    memmap = False
```

Comment:
This definitely needs to be fixed, but I think the issue is with `NDData`.  To be clear, Numpy arrays *do not* participate in cyclic garbage collection.  So if you have a reference cycle somewhere in `NDData` that involves a Numpy array that array will *never* be deallocated (and in the case of arrays with mmap'd buffers, the mmap never closed) unless the reference cycle is explicitly broken.  So that's what you should be looking for.